sdk/lib/atl/atlimage.h

```diff
@@ -93,26 +93,38 @@ class CImage
 
     HDC GetDC() const throw()
     {
-        if (m_hDC)
-            return m_hDC;
+        ATLASSERT(m_nDCRefCount >= 0);
+
+        if (::InterlockedIncrement(&m_nDCRefCount) == 1)
+        {
+            ATLASSERT(m_hDC == NULL);
+            ATLASSERT(m_hOldBitmap == NULL);
+            m_hDC = ::CreateCompatibleDC(NULL);
+            m_hOldBitmap = (HBITMAP)::SelectObject(m_hDC, m_hBitmap);
+        }
 
-        m_hDC = ::CreateCompatibleDC(NULL);
-        m_hOldBitmap = (HBITMAP)::SelectObject(m_hDC, m_hBitmap);
         return m_hDC;
     }
 
     void ReleaseDC() const throw()
     {
-        ATLASSERT(m_hDC);
+        ATLASSERT(m_nDCRefCount > 0);
 
-        if (m_hDC == NULL)
+        if (::InterlockedDecrement(&m_nDCRefCount) != 0)
+            return;
+
+        if (!m_hDC)
```

Comment:
How can `m_hDC` be `NULL`? The existing `ASSERT()` seems right to me.

Reply:
In memory shortage, `CreateCompatibleDC` might fail.